### PR TITLE
HighlightedImage component

### PR DESCRIPTION
**Pull request by aliabid94**
Created a new component called `HighlightedImage`. Allows user to overlay masks on top of an image to highlight regions. Masks can be rectangles (object detection) or 0-1 pixel masks (image segmentation). See demo below for an example:

Recording 2023-04-14 at 12 54 37

Also made some small visual fixes to HighlightedText

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -43,6 +43,26 @@ with gr.Blocks() as demo:
     gr.Video(("video.mp4", "captions.srt"))
 ```
 
+### HighlightedImage Component
```
Comment (abidlabs) [inline CHANGELOG.md:46]: Suggestion that I also put in Slack: how about `AnnotatedImage` component?

File: CHANGELOG.md [hunk 2]
```diff
@@ -43,6 +43,26 @@ with gr.Blocks() as demo:
...
+
+```python
+with gr.Blocks() as demo:
+    img = gr.Image()
+    img_section = gr.HighlightedImage()
+    def mask(img):
+        top_left_corner = [0, 0, 20, 20]
+        random_mask = np.random.randint(0, 2, img.shape[:2])
+        return (img, [(top_left_corner, "left corner"), (random_mask, "random")])
+    img.change(highlight_handler, img, img_section)
```
Comment (freddyaboulton) [inline CHANGELOG.md]: nit: highlight_handler not defined - I think you mean `mask`

File: js/app/src/components/HighlightedImage/HighlightedImage.svelte [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+					/>
+				{/each}
+			</div>
+			{#if show_label}
+				<div class="legend">
+					{#each _value ? _value[1] : [] as [file, label], i}
+						<!-- svelte-ignore a11y-click-events-have-key-events -->
+						<div
+							class="legend-item"
+							style="background-color: {_value
```
Comment (freddyaboulton) [inline js/app/src/components/HighlightedImage/HighlightedImage.svelte]: This is very neat. I think users will want to customize their colors (match branding/theme). I think we should add `color_map`? Or maybe the third element in the tuple is the desired color?

**Comment by abidlabs** [comment]
Really cool tweak to `HighlightedText` @aliabid94!

Recording 2023-04-14 at 11 21 00

**Comment by aliabid94** [comment]
> Really cool tweak to HighlightedText

That's @pngwn's work I believe! I took the inspiration to this component. The fixes to HighlightedText were just some spacing stuff.

File: CHANGELOG.md (hunk 1 above)
Comment (aliabid94) [reply CHANGELOG.md]: Done

File: gradio/components.py [hunk 4]
```diff
@@ -3930,6 +3930,172 @@ def style(
...
+
+    def __init__(
+        self,
+        value: Tuple[
+            np.ndarray | _Image.Image | str,
+            List[Tuple[np.ndarray | Tuple[int, int, int, int], str]],
+        ]
+        | None = None,
+        *,
+        show_legend: bool = True,
```
Comment (abidlabs) [inline gradio/components.py:3951]: I don't think this parameter has any effect. Setting it to `False` still displayed the legend

Comment (aliabid94) [reply gradio/components.py:3951]: fixed

File: gradio/components.py [hunk 5]
```diff
@@ -3930,6 +3930,172 @@ def style(
         return self
 
 
+@document("style")
+class AnnotatedImage(Selectable, IOComponent, JSONSerializable):
```
Comment (abidlabs) [inline gradio/components.py:3934]: Considering adding "Annotatedimage" as an alias for "AnnotatedImage" like we do for "Highlightedtext = HighlightedText" at the bottom of this file

Comment (aliabid94) [reply gradio/components.py:3934]: done

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,7 +3,25 @@
...
+
+```python
+with gr.Blocks() as demo:
+    img = gr.Image()
+    img_section = gr.AnnotatedImage()
+    def mask(img):
+        top_left_corner = [0, 0, 20, 20]
+        random_mask = np.random.randint(0, 2, img.shape[:2])
+        return (img, [(top_left_corner, "left corner"), (random_mask, "random")])
+    img.change(highlight_handler, img, img_section)
```
Comment (abidlabs) [inline CHANGELOG.md:20]:
```suggestion
    img.change(mask, img, img_section)
```

**Review by freddyaboulton — COMMENTED** [review]
This looks very good to me @aliabid94 🔥 

Left a comment about letting users control the colors for the classes.

Also, we need to modify `Dataset` so that this component can be used in an example, no?

**Review by abidlabs — APPROVED** [review]
Really nice component @aliabid94! Tested in Blocks and Interface, as well serialization / deserialization as well and it works too.

cc @hysts this could make for some really cool demos

**Comment by abidlabs** [comment]
> Also, we need to modify Dataset so that this component can be used in an example, no?

I don't think we need to since this is strictly an output component, so it wouldn't ever appear in a `Dataset` (similar to `Label`, for example)

**Comment by freddyaboulton** [comment]
Sounds good - let's not add to dataset unless there is user need! I don't think the distinction between output and input component makes sense though, you can pass any component as input and output to an event handler.

**Comment by freddyaboulton** [comment]
Also, something to keep in mind for 4.0 is giving developers access to pydantic-style data classes for the tuple-typed components, e.g. Code, AnnotatedImage. Dealing with a class is so much nicer than a complex tuple and it makes it easier to add stuff to the interface.

Between these two functions, I think the second is easier to read.  (`img` is supposed to be an AnnotatedImage.)

```python
    def get_selected_label(img, evt: gr.SelectData):
        return [tup[1] for tup in img[1]][evt.index]
    
    def get_selected_label(img: AnnotatedImage, evt: gr.SelectData):
        return [select.label for select in img.selections][evt.index]
```

**Comment by aliabid94** [comment]
Added support for color_map parameter via `.style`! Will merge this in soon if there's not more comments

File: js/app/src/components/HighlightedImage/HighlightedImage.svelte (hunk 3 above)
Comment (aliabid94) [reply js/app/src/components/HighlightedImage/HighlightedImage.svelte]: Done

**Review by freddyaboulton — APPROVED** [review]
You're the bomb @aliabid94 !

**Pull request merged**